### PR TITLE
ensure workflow tries to kill orphan tasks before shutting down

### LINKED ISSUE #4393: cylc stop --kill mode broken?
Is the `--kill` mode working?

Quick example:

```console
$ cylc play <flow>

$ cylc stop --kill <flow>

$ cylc cat-log <flow>
...
2021-09-06T10:52:24Z INFO - Processing 1 queued command(s)
2021-09-06T10:52:24Z INFO - Command succeeded: stop(mode=StopMode.REQUEST_KILL)
2021-09-06T10:52:24Z INFO - Workflow shutting down - REQUEST(KILL)
2021-09-06T10:52:24Z WARNING - Partially satisfied prerequisites left over:
	foo.20191210T0000Z is waiting on:
	* foo_start.20191209T0900Z succeeded
	* cycle_end.20191209T1200Z succeeded
	fool.20191210T0000Z is waiting on:
	* foo_start.20191209T0900Z succeeded
	* cycle_end.20191209T1200Z succeeded
	foot.20191210T0000Z is waiting on:
	* foo_start.20191209T0900Z succeeded
	* cycle_end.20191209T1200Z succeeded
2021-09-06T10:52:24Z WARNING - Orphaned task jobs:
	* bar.20191209T1200Z (running)
	* bool.20191209T1200Z (running)
	* pub.20191209T1200Z (running)
	* foo_start.20191209T0900Z (running)
2021-09-06T10:52:24Z INFO - DONE

$ sqlite3 ~/cylc-run/<flow>/run1/.service/db 'SELECT * FROM task_pool WHERE status = "running"'
20191209T1200Z|bar|p|running|0
20191209T1200Z|bool|p|running|0
20191209T1200Z|pub|p|running|0
20191209T0900Z|foo_start|p|running|0
```

Looks like `--kill` is behaving like `--now`?

Needs a proper test to confirm.

### PR DESCRIPTION
**Pull request by wxtim**
These changes close #4393 

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.py` and
  `conda-environment.yml`.
- [x] Appropriate tests are included (unit and/or functional).
- [x] Appropriate change log entry included.
- [x] No documentation update required.

### CODE REVIEW

File: tests/functional/cylc-stop/00-stop-kill.t [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+init_workflow "${TEST_NAME_BASE}" <<'__FLOW_CONFIG__'
+[scheduling]
+    initial cycle point = 1
+    cycling mode = integer
+    [[graph]]
+        R1 = foo
+[runtime]
+    [[foo]]
+        script = """
+            # Ininite loop:
```
Comment (kinow) [inline tests/functional/cylc-stop/00-stop-kill.t:34]: s/Ininite/Infinite :grimacing:

File: tests/functional/cylc-stop/01-stop-now.t [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+init_workflow "${TEST_NAME_BASE}" <<'__FLOW_CONFIG__'
+[scheduling]
+    initial cycle point = 1
+    cycling mode = integer
+    [[graph]]
+        R1 = foo
+[runtime]
+    [[foo]]
+        script = """
+            # Ininite loop:
```
Comment (kinow) [inline tests/functional/cylc-stop/01-stop-now.t:34]: Ditto :grimacing:

**Review by kinow — APPROVED** [review]
Appears to be waiting for tasks to stop now.

Before:

```bash
2021-09-08T22:22:12+12:00 INFO - Command succeeded:
	stop(mode=StopMode.REQUEST_KILL)
2021-09-08T22:22:12+12:00 INFO - Workflow shutting down - REQUEST(KILL)
2021-09-08T22:22:12+12:00 WARNING - Orphaned task jobs:
	* bar.1 (running)
2021-09-08T22:22:13+12:00 INFO - DONE
```

After change:

```bash
2021-09-08T22:23:43+12:00 INFO - Command succeeded: stop(mode=StopMode.REQUEST_KILL)
2021-09-08T22:23:44+12:00 INFO - [foo.1] status=running (held): (polled)failed at 2021-09-08T22:23:44+12:00  for job(01) flow(r)
2021-09-08T22:23:44+12:00 CRITICAL - [foo.1] -job(01) failed
2021-09-08T22:23:44+12:00 INFO - Workflow shutting down - REQUEST(KILL)
2021-09-08T22:23:45+12:00 INFO - DONE
```

Used the command `cylc stop --kill five` to test it. Looks good @wxtim ! Couple typos, but no blockers. Thanks!!!

File: tests/functional/cylc-stop/00-stop-kill.t (hunk 1 above)
Comment (wxtim) [reply tests/functional/cylc-stop/00-stop-kill.t]:
```suggestion
            # Infinite loop:
```

File: tests/functional/shutdown/21-stop-kill.t [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+waitfor() {
+    # Takes a regex and waits for it to appear in log. Times out after 10
+    # seconds.
+    counter=0
+    # shellcheck disable=SC2143
+    while [[ $(grep -q "${1}" "${WORKFLOW_LOG}") && "${counter}" -lt 10 ]]; do
+        sleep 1
+        counter=$(( counter + 1 ))
+    done
+}
```
Comment (oliver-sanders) [inline tests/functional/shutdown/21-stop-kill.t:55]: Nice, but there is a really simple trick which saves grepping the logs.

```
[runtime]
    [[foo]]
        script = """
            cylc stop --kill "$CYLC_WORKFLOW_NAME"
            while true; do sleep 10; done
        """
        execution time limit = PT1M
```

Then you can

```
cylc play --no-detach "${WORKFLOW_NAME}"
```

File: tests/functional/shutdown/21-stop-kill.t [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+# Wait for foo to start:
+waitfor "foo.1.*started"
+
+run_ok "${TEST_NAME_BASE}" cylc stop --kill "${WORKFLOW_NAME}"
+
+# Wait for scheduler to stop:
+waitfor "INFO - DONE"
+
+grep_ok "foo.1.*failed" "${WORKFLOW_LOG}"
```
Comment (oliver-sanders) [inline tests/functional/shutdown/21-stop-kill.t:65]: This proves that foo failed, however, doesn't prove that it was killed.

I hope we log when we are trying to kill a task, the combination of those two messages should make this watertight.

File: tests/functional/shutdown/22-stop-now.t [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+waitfor() {
+    # Takes a regex and waits for it to appear in log. Times out after 10
+    # seconds.
+    counter=0
+    # shellcheck disable=SC2143
+    while [[ $(grep -q "${1}" "${WORKFLOW_LOG}") && "${counter}" -lt 10 ]]; do
+        sleep 1
+        counter=$(( counter + 1 ))
+    done
+}
```
Comment (oliver-sanders) [inline tests/functional/shutdown/22-stop-now.t:55]: Same trick can be used here.

**Review by oliver-sanders — COMMENTED** [review]
Thanks for the `--now` test.

File: tests/functional/shutdown/22-stop-now.t (hunk 5 above)
Comment (hjoliver) [reply tests/functional/shutdown/22-stop-now.t:55]: If you do need to poll-wait-grep the logs you might have reinvented a wheel @wxtim :
- `poll_grep_workflow_log` - for the test scripts
- `cylc__job__poll_grep_workflow_log` for task job scripts

**Comment by oliver-sanders** [comment]
Shellcheck failures.

**Review by oliver-sanders — APPROVED** [review]

**Comment by oliver-sanders** [comment]
Ach, was waiting for the tests to pass before merging.

I'm afraid `tests/f/shutdown/22-stop-now.t` failed four times in a row (including retries) on Mac OS.

File: tests/functional/shutdown/22-stop-now.t [hunk 6]
```diff
@@ -14,22 +14,34 @@
...
+__FLOW_CONFIG__
+
+run_ok "${TEST_NAME_BASE}-validate" \
+    cylc validate "${WORKFLOW_NAME}"
+
+run_ok "${TEST_NAME_BASE}" cylc play "${WORKFLOW_NAME}" --no-detach --debug
+
+WORKFLOW_LOG="${WORKFLOW_RUN_DIR}/log/workflow/log"
+
+grep_ok 'Orphaned task jobs.*\n.*foo.1' "${WORKFLOW_LOG}" -Pz
```
Comment (oliver-sanders) [inline tests/functional/shutdown/22-stop-now.t:45]: This is the problem, `-P` is not POSIX.

I don't think it is needed here is it?

**Pull request merged**